### ISSUE TITLE
Inconsistent subgradients of matrix frobenius norm

### ISSUE THREAD

**Issue by zjysteven** [pytorch/pytorch#65599]
## 🐛 Bug

I found that the subgradients of matrix frobenius norm (when the norm is 0) are inconsistent across `torch.linalg.norm` and `torch.linalg.matrix_norm`. Specifically, while the former (`torch.linalg.norm`) can give more desired subgradients of 0, the latter (`torch.linalg.matrix_norm`) will produce NaN subgradients. Meanwhile, it is also weird that the results are even inconsistent between `torch.linalg.norm(x)` (default `ord`, which should be frobenius norm for matrix as stated in the documentation) and `torch.linalg.norm(x, ord='fro')`.

## To Reproduce

```python
>>> import torch
>>> a = torch.zeros(3, 3, requires_grad=True)
>>> torch.linalg.matrix_norm(a).backward()
>>> print(a.grad)
tensor([[nan, nan, nan],
        [nan, nan, nan],
        [nan, nan, nan]])
>>> a.grad.zero_();
>>> torch.linalg.norm(a, 'fro').backward()
>>> print(a.grad)
tensor([[nan, nan, nan],
        [nan, nan, nan],
        [nan, nan, nan]])
>>> a.grad.zero_();
>>> torch.linalg.norm(a).backward()
>>> print(a.grad)
tensor([[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]])
```

## Expected behavior

The subgradients should be consistent across different functions that are by definition doing the same thing. And it seems to me that subgradients being 0 in this case is more desirable. Subgradients being NaN simply crashed my training.

## Environment

 - PyTorch Version: 1.9.0+cu111
 - OS: Ubuntu 20.04.1 LTS (x86_64)
 - How you installed PyTorch: pip
 - Build command you used: N/A
 - Python version: 3.8.3
 - CUDA/cuDNN version: 11.1

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @jianyuh @mruberry @walterddr @IvanYashchuk @xwang233

**Comment by mruberry** [comment]
Thanks for reporting this issue, @zjysteven! That is very weird! Looks like there's a bug in the implementation of norm.

**Comment by Lezcano** [comment]
These two norms are different. The frobenius norm is the usual norm given by \norm_{frob}{X} = \sqrt{\sum x_{i,j}^2} = \sqrt{\sum \sigma_i^2} where \sigma_i are the singular values of X. On the other hand, \norm_2(X) is the **operator** 2-norm given by \norm_2{X}= \sigma_0. The fact that these two norms are different explains the different gradients.

Now, \norm_2 is differentiable at every matrix BUT at the matrix zero. Given that the function is non-differentiable, you are getting NaN gradients.

**Comment by Lezcano** [comment]
Nevermind, the default norm in `matrix_norm` is also the Frobenius norm. What I said above is wrong.

The correct thing is that the square root at zero is not differentiable, but we could still take zero as its subgradient of minimum norm. We should correct the formulas to return zero in this case.

**Comment by Jup11ter** [comment]
I have observed this inconsistent gradient (nan and 0) of `torch.linalg.matrix_norm` and `torch.linalg.norm` happened from the 1.9.0 to 1.11.0 version. And in 1.12.0, both of their outputs turn out to be 0.
It seems like this gradient problem is fixed by this [pr #71372](https://github.com/pytorch/pytorch/pull/71372).

In addition, I observe another interesting nan gradient with `torch.linalg.cond` which can be reproduced by enabling the commented codes and happened from 1.9 to 1.13.1.
The latter `a.grad` will always output `NaN`. 

As mentioned above, subgradients being NaN will simply crash the training.
**1) I'm curious if this nan gradient on `torch.linalg.cond` is an expected output.
2) If so, should a note be added to the [docs](https://pytorch.org/docs/stable/generated/torch.linalg.cond.html)? In fact, tensors with 0 value are relatively common, and the NaN gradient it causes will seriously affect subsequent calculations.**

```python
import torch

func_cls=torch.linalg.norm
# func_cls=torch.linalg.cond

a = torch.zeros(3, 3, requires_grad=True)
torch.linalg.matrix_norm(a).backward()
print(a.grad)

a.grad.zero_()
func_cls(a).backward()
print(a.grad)

```

**Comment by lezcano** [comment]
This is fixed in master, they always return zero. This is expected as per point 2 in https://pytorch.org/docs/stable/notes/autograd.html#gradients-for-non-differentiable-functions, given that these functions are convex. This was fixed in https://github.com/pytorch/pytorch/pull/76547

Consider using the nightlies if you are in the chase for bugs :)
